### ISSUE TITLE
Add option to move stashes back to primary sidebar

### ISSUE THREAD

**Issue by LDMFD** [ForkIssues/Tracker#1696]
I'd like them to be visible all the time, so the change in 2.22 where they are in a tab of their own should be optional.

**Comment by jgillard** [comment]
I agree with this. When I saw the release notes I thought they'd move to a box in the bottom-left to separate them from branches, remotes, and tags, but I'm unhappy to see them moved to another tab. The two features in that part of the left sidebar I regularly use are branches and stashes, and now I can't see both at once, so stashes are way more likely to get forgotten about.

**Comment by yossizahn** [comment]
I was also wondering what is the rational behind moving it?
(I don't mind getting used to a different UI but it should have some rational)

**Comment by noheger** [comment]
I would also prefer to have all my stashes directly accessible from the primary sidebar, without having to switch back and forth between two different views.

If I don’t want to see them I can still collapse the Stashes group, so that it then only takes up a single line in the sidebar.

**Comment by deepfriedmind** [comment]
This change doesn't make any sense to me. Please make it optional.

**Comment by AfzalivE** [comment]
May I suggest a middle-ground perhaps.

Keep the stashes in the same tab as branches but in a different section, not part of the same list-tree

This area for example, collapsible and having its own filter input.

<img width="317" alt="Untitled" src="https://user-images.githubusercontent.com/436057/191061300-cde01317-6777-40e3-b5e5-6a718dbb37e8.png">

**Comment by 0xgivn** [comment]
As someone who doesn't look at the toolbar too often, I thought the stashes disappeared altogether.

The previous layout was definitely much more useful to me as I could see them all the time.

**Comment by DanPristupov** [comment]
The recent stashes are already displayed in the commit list and also in the 'stash' dropdown on the toolbar. Making the full list slightly less accessible is OK, I think.

We moved the full stashes list to a separate tab to make them easier to access when there are many branches in the repository. Previously, checking all stashes required either collapsing the branches section or scrolling through it (both of which break focus).

P.S. Such layout is pretty common on macOS. For example in XCode (https://developer.apple.com/xcode/), Nova (https://nova.app/), Calendar (macOS) and many others.

**Comment by koskokos2** [comment]
I don't see much benefit of hiding them either.

Even in larger repos amount of tracked branches should not be too long, so I have a lot of empty space in the sidebar and having stashes hidden, while most of the sidebar is empty feels limiting.

In the same time commit list is always very busy, so it is not easy to look stashes in there.

**Comment by deepfriedmind** [comment]
Here I have a decent amount of local and remote branches and it still only takes up 50% of the sidebar:

![frontend-desktop (Fork) - 2022-09-20 at 13 35 39](https://user-images.githubusercontent.com/1065569/191249103-90e814d7-eb64-41d8-8db8-6416d91692af.png)

**Comment by bradser** [comment]
I happen to have a large number of stashes. The dropdown limits to 15. How do I see all my stashes now?

**Comment by DanPristupov** [comment]
@bradser open the stashes section on the sidebar.

**Comment by noheger** [comment]
I’m working with the new Stashes layout for a couple of days now. The constant switching back and forth between these two separate sidebar views is really annoying and time consuming. For me this is a huge regression.

I understand that for some users it might be handy to have stashes in a separate view if they also have lots of branches shown in the sidebar (which I don’t).

So why not just showing stashes in both places? In a separate view for those who work with many branches, but also in the main view for those who don’t? It only takes up one additional row in the primary sidebar when it’s closed, so it doesn’t consume much screen estate anyway.

Or make this additional entry in the primary sidebar optional, like the OP suggested.

**Comment by n8gray** [comment]
Just as a counterpoint, I really like the new layout. Having stashes mixed in with a million branches was always a pain for me (my company has insane amounts of branches).

**Comment by LDMFD** [comment]
Indeed, please note I never asked for the new layout to be removed - only to be complemented by the old in a switchable setting.

Fwiw, I actually consider stashes a form of tech debt, as they are kept locally and not 'backed up' by pushing a branch, therefore keeping them up-front is a constant reminder to try to prune the list as I go.

**Comment by sanderschnydrig** [comment]
Thankfully I stumbled upon this thread or I would still be wondering where my stashed went. I appreciate the layout the way it is now but I was not able to figure out what happened only casually using the app. My thinking was they were moved next to the push button.

**Comment by kuthair** [comment]
It's annoying because I can't find and delete all my stashes in one spot, I have to go hunting and pecking thru the commit tree to find them.

**Comment by DanPristupov** [comment]
@kuthair stashes are in a separate tab on the sidebar

**Comment by ShlomoCode** [comment]
@fork-dev Plan to make it optional?
It seems to be important to a lot of people (including me 😀 )

**Comment by kuthair** [comment]
> @kuthair stashes are in a separate tab on the sidebar

Oh! There it is... that works for me, thank you.

This may actually solve a lot of people's problems... Maybe they missed that icon too?

**Comment by DanPristupov** [comment]
@ShlomoCode it will definitely not be optional. We may either keep it or roll back. Currently we are trying to figure out which layout is better.

@kuthair

> Maybe they missed that icon too?

Yes, probably. We mentioned that in the release notes though.

**Comment by kuthair** [comment]
I actually like it, now that I see it. Sorry, I didn't read the release notes.

**Comment by LDMFD** [comment]
@DanPristupov just curious, what is the rationale behind rejecting the original proposal (to add an option to choose from either)?

I appreciate it is an overhead to maintain both, if this is the driving factor.

**Comment by DanPristupov** [comment]
@LDMFD 

There are two problems.

1. The complexity of having a duplicated layout (and functionality).

- Copy-pasting stash codebase to both places is a bad idea
- Extracting stash GUI as an abstract entity which can be displayed in different layouts and handle context menu, filtering, section headers (treeview parent vs title), dynamic switching etc would take weeks.

2. 99.5% of users never change default settings. So, introducing an option won't make them happier, it would help just a minor group of people.

As for me, the new layout has both pros and cons. It is more clean, but also adds additional nesting and complexity.

If new layout is worse than the original one, we should roll it back. If it's better, then we should keep it. Introducing an option doesn't make much sense, because people won't change it anyway.

**Comment by DanPristupov** [comment]
It looks like the new layout adds unnecessary complexity to the UI.

The feature needs to be reimagined and we will likely roll the layout back until it's polished out (or we find a better approach).

**Comment by LDMFD** [comment]
Dan, thanks for both giving your perspective on the original proposal (all reasonable), and taking into account the feedback given.

**Comment by grigorye** [comment]
Personally, I *really* loved the stashes that had the dedicated button - I thought that was the right design change, I was really happy when it was introduced. Now I miss it after upgrading to 2.23.1. My reasoning is that stashes are quite different from the rest of the entities in that tree - they exist for different purposes and are used in different workflows (ok, that's my personal perspective). I don't get why I should scroll or collapse branches or bother about the rest of the tree at all, when I just want to see the stashes. Overall, going through a scrollable UI for even accessing the stashes is (imho) much worse design choice/UX than just having a dedicated toggle at a fixed position.

P.S. Not everybody has just a bunch of branches. Not everybody has lots of vertical space. I'm working on a project where I have hundreds of branches and hundreds of tags. It's *really* hard to navigate to stashes in this case, if you ever open branches.

**Comment by grigorye** [comment]
Btw, I sympathize the people that wanted to bring stashes back alongside the branches, and appreciate the fact that they find it "less confusing", but I just can't explain it, why it is so useful - I mean, "Stashes" is collapsed by default anyway, so unless we expand it we can't see any of the stashes. Then, how that click to expand (or collapse) on a that (scrollable) item in the list, is better or easier than *clicking* on a fixed button in the toolbar?

P.S. I'm not sure about it, but I feel like people just get used to it (especially when it did not bother them before) and then (of course) get annoyed when it was "gone"/UI was changed. However, I have some doubts that a separate "Stashes" button would be a problem for the *new* users (leaving aside people who actually had problems with the original/now current UX).

**Comment by LDMFD** [comment]
The new layout wasn't confusing (other than discovery) - it's just more opinionated. I think the advantage of keeping them all as root nodes in a single tree is that it adapts more easily to the multitude of personal workflows that exist.

E.g. you can collapse the other root nodes to see Stashes in isolation (fig. 1). The sub-tree states are preserved when each is re-opened. Also, use `/` in tag and branch names to create sub-folders (fig. 2).

<img width="222" alt="image" src="https://user-images.githubusercontent.com/1093972/197901447-fc5f2a74-f5c0-44c2-8d3a-743a9bfa70f2.png">

<img width="183" alt="image" src="https://user-images.githubusercontent.com/1093972/197901593-8c92b06e-218f-437a-b922-90e882fb9a5f.png">

**Comment by grigorye** [comment]
> The new layout wasn't confusing (other than discovery) - it's just more opinionated. I think the advantage of keeping them all as root nodes in a single tree is that it adapts more easily to the multitude of personal workflows that exist.

Well... From my perspective, it does *not* "adapt" to the case of hundreds of items - and that's not an "opinion", rather a real usability problem, unfortunately.

As for "more opinionated" - I feel actually that the "tree" is not less opinionated - I mean, I still don't get why stashes and the rest (branches/tags) are put together at all from the beginning, so that interacting with one kind of items affects interaction with another kinds of items. I mean, unfortunately, I could not see any explanation for that in the above comments/likes, other than "discoverability" (that's indeed a valid point, provided that there was just an icon to denote stashes vs "text" labels for the rest).

> E.g. you can collapse the other root nodes to see Stashes in isolation (fig. 1).

Yeah, I get that. The question is how to collapse/expand those root nodes if I have hundreds of items in both branches and stashes.

> The sub-tree states are preserved when each is re-opened. Also, use `/` in tag and branch names to create sub-folders (fig. 2).

Well, we use (standard) git-flow, and we have feature/ release/ bug/ subfolders (btw, we have sub-subfolders inside too). It does *not* help at all (we still have hundred of items under subfolders). I have to "collapse" branches (leaving aside scrolling to that root item if I'm not at the beginning of the list, just to collapse it), then "expand" stashes, *every time* (ok, I could try keeping stashes always open to save the number of clicks, but I'm not sure that I do that). *It's twice the number of clicks + scrolling*. Btw, it does not help with "always seeing stashes" either because after expanding the branches, stashes are scrolled out.

Long story short, in my opinion, the old/current setup *works* relatively well only in a very limited number of cases when the whole list fits (vertically) into the view when expanded.

**Comment by n8gray** [comment]
I just have to disagree with the reason stated for reverting this improvement, that the new design added "unnecessary complexity to the UI". The new layout was not in any way more complex. It used a simple pattern (tabs) we're all familiar with from dozens of apps we use every day. It allowed us to focus on stashes instead of mentally filtering out the visual noise from branches, tags, remotes, etc, which was a real improvement for those of us who work in repos with hundreds or thousands of branches and tags. 

Collapsing and then reopening all the roots in the sidebar is a clunky multi-click workaround. Need to use search to find a specific stash? Oops, all those roots are open again showing search hits, and your stashes are now pushed off the bottom of the view!

So what we've got is a tangible step backwards for "big repo" people, and for what? People who like to have their stashes and branches both open at once, and have the luxury of a small repo where that's feasible? Or was it just "they moved my cheese?" From what I can tell, a significant fraction of the comments on this thread were from people who didn't realize the stashes had moved, and were pretty much fine when they understood the new layout. 

I doubt that "reimagining" the feature will get you very far because it seems like the real problem wasn't the new layout, it was onboarding people into the new layout.

**Comment by ShlomoCode** [comment]
I think that most of the users of "fork"/most repositories, are not repositories of hundreds/thousands of **active** branches. Look at GitHub - how many such repositories are there?
Of course, the number of branches in the cloud is irrelevant, you can permanently shrink the Remotes section. And who needs to go regularly on their local computer between exactly hundreds of branches??

**Comment by n8gray** [comment]
There are plenty of git clients that work fine with small repos. That's easy. The reason I paid real money for Fork is that it is one of the very few that handles big repos well. If the plan is to ignore the needs of serious work repos and focus on hobby github repos then I'll be moving on.

> Of course, the number of branches in the cloud is irrelevant, you can permanently shrink the Remotes section. And who needs to go regularly on their local computer between exactly hundreds of branches??

Nope, the moment you use search the hundreds of remote branches pop open with hundreds of results.

**Comment by ShlomoCode** [comment]
> Nope, the moment you use search the hundreds of remote branches pop open with hundreds of results.

How is this related? The search is already in a separate tab...

<img width="307" alt="CleanShot 2022-10-26 at 22 16 08" src="https://user-images.githubusercontent.com/78599753/198116566-c570cfde-bb9e-4e59-abf0-2d0a260441ff.png">

**Comment by n8gray** [comment]
Sorry, I meant the "Filter" feature, not "Search".

**Comment by n8gray** [comment]
<img width="347" alt="Screen Shot 2022-10-26 at 12 27 50 PM" src="https://user-images.githubusercontent.com/82794/198118628-f5bbbce2-f885-4ce5-8a3f-483b776c71c7.png">

**Comment by bradser** [comment]
Yes it was a 'moved cheese' situation for me, happy either way.

**Comment by deepfriedmind** [comment]
Thank you for listening to the feedback and changing it back!

**Comment by grigorye** [comment]
> I think that most of the users of "fork"/most repositories, are not repositories of hundreds/thousands of **active** branches.

How can we know that?

> Look at GitHub - how many such repositories are there? 

Unfortunately, I don't know. Do you have an answer to that question, based on some data?

Overall, I'm not sure if that is a good example for this discussion:

- Public GitHub (that we can observe) is just an "iceberg".
- Local branches are not necessarily *active* branches.
- Fork users are not necessarily public GitHub users.
- Git is a *way* bigger concept than GitHub, Fork is not GitHub-only client.

Btw, I wonder if @DanPristupov has any real statistics regarding the usage of branches/types of users and etc. in the client, that you could share.

> Of course, the number of branches in the cloud is irrelevant, you can permanently shrink the Remotes section. And who needs to go regularly on their local computer between exactly hundreds of branches??

Re: local branches. It's not necessarily about going/checking or working with those branches simultaneously. They just *can* already be there, for instance as a result of working on them at some time. Such things just naturally happen, and e.g. accumulate over time. It's not a big deal to have them, given that Git branches are cheap. I can say that I switch around 10 *new* branches each week - it just happens because there're several people (around 5-7) working on the project, creating branches regularly. Some branches are mine, some branches come from other people, mainly for review. Yep, most of those branches become “stale” over time/I don't return to most of them (things happen though, from time to time), but it did not bother me at all, until I was forced to deal with scrolling/collapsing and etc.

**Comment by pionxzh** [comment]
I know moving it to a separate tab is a reasonable move. But I still hope that we can have the option to bring it back.
To be honest, I turned off the automatic updates after seeing the release note.
Maybe it's just because I'm a creature of habit.

**Comment by Pavel-Laukhin** [comment]
Opposite to pionxzh, to be honest, I turned off the automatic updates after seeing the release note when stashes turned from the separate tap back to branches and tags. What a mess it would be! Again. Infinite scroll from top to bottom over and over again... What a beautiful decision it was when stashes moved to the separate tab! So sad that it was turned back to the previous UX. 😭
